Consider checking if git submodules have been initialized

We use git submodules that require `git submodule update --init` for successful builds. A number of users have encountered issues after checking out the project or changing to a commit that includes new submodules. One idea is to warn about uninitialized submodules in CMake. Dirty submodules (submodules with local changes) are fine and not worth warning about, but empty submodules (uninitialized) are not.

There is a suggestion at https://cmake.org/pipermail/cmake/2012-May/050190.html:

```cmake
if( NOT EXISTS "${MY_SOURCE_DIR}/project/.git" )
message( SEND_ERROR "The git submodules are not available. Please run
git submodule update --init --recursive"
)
endif()
```

We could do something similar, iterating over all submodules we expect to be initialized and recommending `git submodule update --init` (no `--recursive`).

Sure!

We have a number of dependencies in IREE that are included via git submodules ([docs](https://git-scm.com/docs/git-submodule)). These are all enumerated in [.gitmodules](https://github.com/google/iree/blob/main/.gitmodules) and are located in [third_party/](https://github.com/google/iree/tree/main/third_party). A `git clone` or `git checkout` won't "initialize" those submodules, so they'll be empty directories.

If someone tries to build without a submodule initialized, they'll get an error like this one:

```
CMake Error at build_tools/cmake/external_cc_library.cmake:108 (add_library):
  Cannot find source file:

    .../iree/third_party/libyaml/src/api.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  build_tools/third_party/libyaml/CMakeLists.txt:9 (external_cc_library)


CMake Error at build_tools/cmake/external_cc_library.cmake:108 (add_library):
  No SOURCES given to target: yaml_yaml
Call Stack (most recent call first):
  build_tools/third_party/libyaml/CMakeLists.txt:9 (external_cc_library)
```

that error message sort of explains what happened (a file is missing), but we could provide a more obvious message in the case that a submodule is not initialized at all, like

```
git submodule "third_party/libyaml" is not initialized, run "git submodule update --init"
```

To check for uninitialized submodules and print such an error message, we would put some CMake code around here: https://github.com/google/iree/blob/790ed1ad9b8486258b112979ac2c83ef33708f85/CMakeLists.txt#L263-L265

maybe referencing a helper function defined in [build_tools/cmake/iree_macros.cmake](https://github.com/google/iree/blob/main/build_tools/cmake/iree_macros.cmake) (to keep the root CMakeLists.txt less complicated).

That code would go through directories in `third_party/` checking for uninitialized submodules in some way (e.g. checking for empty directories, checking for paths from `.gitmodules` that don't yet have a `.git` file, or running some command(s) from https://git-scm.com/docs/git-submodule). If an uninitialized submodule is found, that code would print a fatal error message. If all submodules are initialized, the code should not print anything.

CMake use is generally split into a _configure_ step and a _build_ step. We want to check for uninitialized submodules at configure time, before developers try to build.

Some CMake like this around that location would let you run your script and check its result:

```cmake
#-------------------------------------------------------------------------------
# Check if git submodules have been initialized
#-------------------------------------------------------------------------------

execute_process(
    COMMAND false # (your command would go here)
    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
    RESULT_VARIABLE ret
)
if(NOT ret EQUAL "0")
  message(FATAL_ERROR "Command failed: ${ret}")
else()
  message(STATUS "Success")
endif()
```

We have instructions on building IREE here: https://google.github.io/iree/building-from-source/getting-started/#clone-and-build (see also [general documentation here](https://cmake.org/runningcmake/)), and the `cmake -B ../iree-build/ -DCMAKE_BUILD_TYPE=RelWithDebInfo .` configure command is where that code would run and pass/fail, with output like

```
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.exe" --no-warn-unused-cli -DIREE_TARGET_BACKENDS_TO_BUILD=all -DIREE_HAL_DRIVERS_TO_BUILD=all -DIREE_BUILD_TESTS=ON -DIREE_BUILD_COMPILER=ON -DIREE_ENABLE_MLIR=ON -DIREE_BUILD_PYTHON_BINDINGS=OFF -DIREE_BUILD_SAMPLES=ON -DIREE_ENABLE_RUNTIME_TRACING=ON -DCMAKE_TOOLCHAIN_FILE=D:/dev/tools/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_INSTALL_PREFIX=d:/dev/projects/iree/../iree-build/install -Wno-dev -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -Hd:/dev/projects/iree -Bd:/dev/projects/iree-build -G Ninja
[cmake] Not searching for unused variables given on the command line.
[cmake] -- Building HAL drivers: Cuda;DyLib;VMVX;Vulkan
[cmake] -- Building target backends: CUDA;DYLIB-LLVM-AOT;WASM-LLVM-AOT;Metal-SPIRV;ROCM;Vulkan-SPIRV;VMVX
[cmake] -- Found Python3: C:/Python38/python.exe (found version "3.8.2") found components: Interpreter 
[cmake] CMake Error at CMakeLists.txt:334 (message):
[cmake]   Command failed: 1
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "D:/dev/projects/iree-build/CMakeFiles/CMakeOutput.log".
[cmake] See also "D:/dev/projects/iree-build/CMakeFiles/CMakeError.log".
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -329,7 +329,7 @@ endif()
 find_package(Python3 COMPONENTS Interpreter QUIET)
 if(Python3_FOUND)
   execute_process(
-    COMMAND python scripts/git/check_submodule_init.py
+    COMMAND python3 scripts/git/check_submodule_init.py
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     RESULT_VARIABLE ret
   )

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -322,6 +322,20 @@ if(${IREE_BUILD_COMPILER} OR
 endif()
 
 #-------------------------------------------------------------------------------
+# Check if git submodules have been initialized.
+# This will only run if python/python3 is available
+#-------------------------------------------------------------------------------
+
+execute_process(
+  COMMAND python --version && python scripts/git/check_submodule_init.py
+  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+  RESULT_VARIABLE ret
+)
+if(NOT ret EQUAL "0")
+  message(FATAL_ERROR "check_submodule_init.py failed with code: ${ret}")
+endif()
+
+#-------------------------------------------------------------------------------
 # MLIR/LLVM Dependency
 # We treat the LLVM dependency specially because we support several different
 # ways to use it:

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -327,7 +327,7 @@ endif()
 #-------------------------------------------------------------------------------
 
 execute_process(
-  COMMAND python --version && python scripts/git/check_submodule_init.py
+  COMMAND !(python --version) || python scripts/git/check_submodule_init.py
   WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   RESULT_VARIABLE ret
 )

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -329,7 +329,7 @@ endif()
 find_package(Python3 COMPONENTS Interpreter QUIET)
 if(Python3_FOUND)
   execute_process(
-    COMMAND python3 scripts/git/check_submodule_init.py
+    COMMAND ${Python3_EXECUTABLE} scripts/git/check_submodule_init.py
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     RESULT_VARIABLE ret
   )

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -323,16 +323,19 @@ endif()
 
 #-------------------------------------------------------------------------------
 # Check if git submodules have been initialized.
-# This will only run if python/python3 is available
+# This will only run if python3 is available.
 #-------------------------------------------------------------------------------
 
-execute_process(
-  COMMAND !(python --version) || python scripts/git/check_submodule_init.py
-  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-  RESULT_VARIABLE ret
-)
-if(NOT ret EQUAL "0")
-  message(FATAL_ERROR "check_submodule_init.py failed with code: ${ret}")
+find_package(Python3 COMPONENTS Interpreter QUIET)
+if(Python3_FOUND)
+  execute_process(
+    COMMAND python scripts/git/check_submodule_init.py
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    RESULT_VARIABLE ret
+  )
+  if(NOT ret EQUAL "0")
+    message(FATAL_ERROR "check_submodule_init.py failed, see the logs above")
+  endif()
 endif()
 
 #-------------------------------------------------------------------------------

--- a/scripts/git/check_submodule_init.py
+++ b/scripts/git/check_submodule_init.py
@@ -5,27 +5,28 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 import os
+import subprocess
 import sys
 
 
 def run():
-  # Checking the "git status"
-  git_check = os.popen("git status")
-  git_status = git_check.readline()
+  # No-op if we're not in a git repository.
+  try:
+    subprocess.check_call(['git', 'rev-parse', '--is-inside-work-tree'],
+                          stdout=subprocess.DEVNULL,
+                          stderr=subprocess.DEVNULL)
+  except:
+    return
 
-  # no-op if not a git repository
-  if not git_status.startswith("On"):
-    sys.exit(1)
-  else:
-    output = os.popen("git submodule status")
-    submodules = output.readlines()
+  output = os.popen("git submodule status")
+  submodules = output.readlines()
 
-    for submodule in submodules:
-      if (submodule.strip()[0] == "-"):
-        print(
-            f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`"
-        )
-        sys.exit(1)
+  for submodule in submodules:
+    if (submodule.strip()[0] == "-"):
+      print(
+          f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`"
+      )
+      sys.exit(1)
 
 
 if __name__ == "__main__":

--- a/scripts/git/check_submodule_init.py
+++ b/scripts/git/check_submodule_init.py
@@ -9,22 +9,24 @@ import sys
 
 
 def run():
-    # Checking the "git status"
-    git_check = os.popen("git status")
-    git_status = git_check.readline()
+  # Checking the "git status"
+  git_check = os.popen("git status")
+  git_status = git_check.readline()
 
-    # no-op if not a git repository
-    if (git_status.startswith("On")):
-        output = os.popen("git submodule status")
-        submodules = output.readlines()
+  # no-op if not a git repository
+  if not git_status.startswith("On"):
+    sys.exit(1)
+  else:
+    output = os.popen("git submodule status")
+    submodules = output.readlines()
 
-        for submodule in submodules:
-            if (submodule.strip()[0] == "-"):
-                print(
-                    f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`")
-
-            sys.exit(1)
+    for submodule in submodules:
+      if (submodule.strip()[0] == "-"):
+        print(
+            f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`"
+        )
+        sys.exit(1)
 
 
 if __name__ == "__main__":
-    run()
+  run()

--- a/scripts/git/check_submodule_init.py
+++ b/scripts/git/check_submodule_init.py
@@ -1,0 +1,30 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import os
+import sys
+
+
+def run():
+    # Checking the "git status"
+    git_check = os.popen("git status")
+    git_status = git_check.readline()
+
+    # no-op if not a git repository
+    if (git_status.startswith("On")):
+        output = os.popen("git submodule status")
+        submodules = output.readlines()
+
+        for submodule in submodules:
+            if (submodule.strip()[0] == "-"):
+                print(
+                    f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`")
+
+            sys.exit(1)
+
+
+if __name__ == "__main__":
+    run()

--- a/scripts/git/check_submodule_init.py
+++ b/scripts/git/check_submodule_init.py
@@ -24,8 +24,8 @@ def run():
   for submodule in submodules:
     if (submodule.strip()[0] == "-"):
       print(
-          f"The git submodule '{submodule.split()[1]}' is not initialized. Please run `git submodule update --init`"
-      )
+          "The git submodule '%s' is not initialized. Please run `git submodule update --init`"
+          % (submodule.split()[1]))
       sys.exit(1)
 
 

